runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +150,146 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
+ */
+jstring JNICALL
+Java_openj9_tools_attach_diagnostics_base_DiagnosticUtils_getHeapClassStatisticsImpl(JNIEnv * env, jclass unused)
+{
+	J9VMThread *vmThread = (J9VMThread *) env;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9HashTable *statsTable = NULL;
+	J9HashTableState hashTableState;
+	BOOLEAN outOfMemory = FALSE;
+	j9object_t stringObject = NULL;
+	J9HeapStatisticsTableEntry **statsArray = NULL;
+	jstring stringObjectRef = NULL;
+	I_32 numClasses = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	vmFuncs->internalEnterVMFromJNI(vmThread);
+
+	vmFuncs->acquireExclusiveVMAccess(vmThread);
+	statsTable = collectHeapStatistics(vmThread);
+	vmFuncs->releaseExclusiveVMAccess(vmThread);
+
+	if (NULL != statsTable) {
+		numClasses = hashTableGetCount(statsTable);
+		statsArray = j9mem_allocate_memory(numClasses * sizeof(J9HeapStatisticsTableEntry*), J9MEM_CATEGORY_VM_JCL);
+	}
+	if (NULL == statsArray) {
+		outOfMemory = TRUE;
+	} else {
+		I_32 cursor = 0;
+		J9HeapStatisticsTableEntry *entry = (J9HeapStatisticsTableEntry *) hashTableStartDo(statsTable, &hashTableState);
+		while (NULL != entry) {
+			entry->aggregateSize = entry->objectSize * entry->objectCount;
+			statsArray[cursor] = entry;
+			cursor += 1;
+			entry = (J9HeapStatisticsTableEntry *) hashTableNextDo(&hashTableState);
+		}
+		hashTableFree(statsTable);
+		numClasses = cursor; /* adjust the length in case the hash table contained nulls */
+		sortStatsTable(statsArray, 0, numClasses - 1);
+	}
+	if (!outOfMemory) {
+		I_32 printedLength = 0;
+		UDATA bufferSize = 0;
+		do {
+			bufferSize += numClasses * 80; /* try incrementally larger sizes */
+			char *stringBuffer = (char *) j9mem_allocate_memory(bufferSize, J9MEM_CATEGORY_VM_JCL);
+			if (NULL == stringBuffer) {
+				outOfMemory = TRUE;
+				break;
+			}
+			printedLength = printHeapStatistics(env, statsArray,numClasses, stringBuffer, bufferSize);
+			if (printedLength > 0) {
+				stringObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread,
+						(U_8 *) stringBuffer, printedLength, J9_STR_XLAT);
+				stringObjectRef = vmFuncs->j9jni_createLocalRef(env, stringObject);
+			}
+			j9mem_free_memory(stringBuffer);
+		} while (0 == printedLength);
+		j9mem_free_memory(statsArray);
+	}
+
+	if (outOfMemory) {
+		Trc_JCL_heapStatisticsOOM(vmThread);
+		vm->internalVMFunctions->setNativeOutOfMemoryError(vmThread, 0, 0);
+	}
+	vm->internalVMFunctions->internalExitVMToJNI(vmThread);
+
+	return stringObjectRef;
+}
+
+static I_32
+printHeapStatistics(JNIEnv * env,J9HeapStatisticsTableEntry **statsArray,
+		U_32 numClasses, char *stringBuffer, UDATA bufferSize)
+{
+	char *bufferCursor = stringBuffer;
+	I_32 classCursor = 0;
+	UDATA cumulativeCount = 0;
+	UDATA cumulativeSize = 0;
+	UDATA result = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+			"%5s%15s%15s    %s\n-------------------------------------------------\n",
```

Comment:
This format string should have the same structure as the one below: i.e. `"%5s %14s %14s ..."`.